util/util.go

```diff
@@ -193,29 +195,42 @@ func ObjectKey(object metav1.Object) client.ObjectKey {
 
 // ClusterToInfrastructureMapFunc returns a handler.ToRequestsFunc that watches for
 // Cluster events and returns reconciliation requests for an infrastructure provider object.
-func ClusterToInfrastructureMapFunc(gvk schema.GroupVersionKind) handler.MapFunc {
+func ClusterToInfrastructureMapFunc(ctx context.Context, gvk schema.GroupVersionKind, c client.Client, providerCluster client.Object) handler.MapFunc {
+	log := ctrl.LoggerFrom(ctx)
 	return func(o client.Object) []reconcile.Request {
-		c, ok := o.(*clusterv1.Cluster)
+		cluster, ok := o.(*clusterv1.Cluster)
 		if !ok {
 			return nil
 		}
 
 		// Return early if the InfrastructureRef is nil.
-		if c.Spec.InfrastructureRef == nil {
+		if cluster.Spec.InfrastructureRef == nil {
 			return nil
 		}
 		gk := gvk.GroupKind()
 		// Return early if the GroupKind doesn't match what we expect.
-		infraGK := c.Spec.InfrastructureRef.GroupVersionKind().GroupKind()
+		infraGK := cluster.Spec.InfrastructureRef.GroupVersionKind().GroupKind()
 		if gk != infraGK {
 			return nil
 		}
 
+		providerCluster := providerCluster.DeepCopyObject().(client.Object)
+		key := types.NamespacedName{Namespace: cluster.Spec.InfrastructureRef.Namespace, Name: cluster.Spec.InfrastructureRef.Name}
+		if err := c.Get(ctx, key, providerCluster); err != nil {
```

Comment:
I'm not sure if it's correct to use the ctx from startup for this call.

Reply:
Which one would you use instead? That one usually denotes the liftetime of the program so that doesn't seem wrong?

Reply:
Agree, there isn't a better context available. 

I saw us use `context.TODO` in other cases (at least `ClusterToObjectsMapper`). Not sure, but it might be ideal to use the one from the current "Reconcile" eventually but that is not supported today by the `MapFunc` interface.

So it might be a question of if we think that we might get a better context there eventually.

No strong opinion either way, but if possible I would like to do it consistently across our MapFunc implementations.

Reply:
(cc @vincepri)